### ISSUE TITLE
Display keeps blinking when program get error

### ISSUE THREAD

**Issue by jneb** [zathras/jrpn#108]
JRPN15C on WIndows

When the solvers can't find a zero of a function, "Error 8" appears, but the display stays blinking. There is no easy way to turn this off.

**Debug Log**
{"version":1,"modelName":"15C","settings":{"menuEnabled":true,"windowEnabled":false,"longNumbers":1,"euroComma":true,"showAccelerators":true,"systemOverlaysDisabled":false,"orientation":0,"keyFeedback":0,"msPerInstruction":0.0},"displayMode":"x6","integerSignMode":"2","wordSize":56,"stack":["1000000000000","0","9999999999099","1000000000000"],"lastX":"1000000000000","flags":[false,false,false,false,false,false,false,false,false,true],"memory":{"storage":"20b61001867801861978b71061018676000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000002000000000000","program":{"lines":16,"currentLine":0,"returnStack":[0,0,0,0,0,0,0],"returnStackPos":-1},"I":"0"},"debugLog":{"initialState":{"version":1,"modelName":"15C","settings":{"menuEnabled":true,"windowEnabled":false,"longNumbers":1,"euroComma":true,"showAccelerators":true,"systemOverlaysDisabled":false,"orientation":0,"keyFeedback":0,"msPerInstruction":0.0},"displayMode":"x6","integerSignMode":"2","wordSize":56,"stack":["1000000000000","0","9999999999099","1000000000000"],"lastX":"1000000000000","flags":[false,false,false,false,false,false,false,false,false,true],"memory":{"storage":"20b61001867801861978b71061018676000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000002000000000000","program":{"lines":16,"currentLine":0,"returnStack":[0,0,0,0,0,0,0],"returnStackPos":-1},"I":"0"},"numRegisters":20,"resultMatrix":0,"matrices":[{"columns":0,"values":[],"rowSwaps":null},{"columns":0,"values":[],"rowSwaps":null},{"columns":0,"values":[],"rowSwaps":null},{"columns":0,"values":[],"rowSwaps":null},{"columns":0,"values":[],"rowSwaps":null}],"lastRandom":0.27765501041926766},"keys":[-1,43,0]},"endDisplay":"   error 8  ","numRegisters":20,"resultMatrix":0,"matrices":[{"columns":0,"values":[],"rowSwaps":null},{"columns":0,"values":[],"rowSwaps":null},{"columns":0,"values":[],"rowSwaps":null},{"columns":0,"values":[],"rowSwaps":null},{"columns":0,"values":[],"rowSwaps":null}],"lastRandom":0.27765501041926766}

**To Reproduce**
I wrote a function that gives an error when evaluated at 1 (let's say 1 - / ). Then I solved it with 1 as a given value.

**Expected behavior**
Error 8 (what it does), but not blinking.

**Platform:**
Windows 10

**Comment by zathras** [comment]
To stop blinking, g-CF 9 is the standard key sequence.

Instructions to reproduce are incomplete, but with this program:
```
#  Program produced by JRPN 15C.
#  Generated 2024-7-2 11:06 PDT.
#  Program occupies 7 bytes.

   000 {          } 
   001 { 42 21 11 } f LBL A
   002 {        0 } 0
   003 {       10 } /

# End.

```
and then `1 enter 2 f solve A`, my real 15C gives Error 0 (not blinking), and the simulator gives a blinking Error 8.  Playing around with the real calculator a bit, it looks like when there's an error in the program, it immediately stops trying to solve, and shows the error from the program, not Error 8.  Error 8 seems to be reserved for when it fails to find a solution, but the program it runs doesn't encounter an error.

**Comment by zathras** [comment]
I checked the real 15C, and an Error 0 in a called subroutine during solve or integrate halts the solve/integrate  attempt at the line that generates the error.  But `runSubroutine` (runners.dart line 131) introduces `okErrors`, which claims that it's OK to get an Error 0 or an Error 1 in a called subroutine.  Git blame takes me back to https://github.com/zathras/jrpn/commit/50825047008c041b455e6f94a0cdac0e636aa208 .  Hmmm....  No issue attached to that, and this part of the code was under active development at the time.  It mentions sqrt(-1) as the justification.

I tried solving sqrt(x - 1) between .99999 and 2 on the real 15C, and it also halts on Error 0.  So, the commit messages justification doesn't match the behavior of the calculator.

However, there is one case where errors should be ignored:  The integration algorithm I use initially tries to evaluate the endpoints (unlike the real calculators).  If either endpoint fails, it falls back to a different algorithm.  So, during the evaluation of integration endpoints, I should tolerate an Error 0.  It looks like that's the only time I should, though; in other cases, solve and integrate should terminate.

I'm not quite sure why it's flashing Error 8.  When solve fails, both the simulator and the real calculator give a steady Error 8,

**Comment by zathras** [comment]
About overflow:  solve on the real calculator doesn't clear the condition.  When I solve (x + 1) ^ 99 ^ 99 between 0 and 1, solve finds the root at -1, but along the way the overflow flag gets set, and not cleared.

**Comment by zathras** [comment]
OK, I see why I had `okErrors`.  "Users/Eddie Shore/Reactance chart solver.15c" gives an error 0 when my solve algorithm goes outside the range [0,1].  The provided function gives Error 0 when the argument is < 0.  Looking at what the real 15C does, it tries harder to keep inside the desired range, whereas my algorithm goes outside that range on just the fourth try.  Back when I was writing this, observing the Error 0 mislead me into thinking that the calculator was ignoring errors in the evaluated function, but it wasn't.  It just wasn't triggering an error in this  particular case.

So I think the thing to do is to be faithful to the original, and propagate errors out, BUT tweak the solve function to stick within the provided range, until some heuristic point for giving up is reached.  Unless I'm willing to copy the original's algorithm (and break cleanroom status), that's about the best I can do.

Of the regression tests, this is the only one that fails when I get rid of `okErrors`, FWIW.

**Comment by zathras** [comment]
@jneb Would you care to let me know how to reproduce your problem?  In order to get the error, what range did you give the solve key?  What is the expected root?

From the memory dumb, I was able to recover this program listing:
```
#  Program produced by JRPN 15C.
#  Generated 2024-7-15 19:41 PDT.
#  Program occupies 21 bytes.

   000 {          } 
   001 { 42 21  0 } f LBL 0
   002 {    45  0 } RCL 0
   003 {       14 } yX
   004 {        1 } 1
   005 {       30 } -
   006 {    43 33 } g R^
   007 {        1 } 1
   008 {       30 } -
   009 {       10 } /
   010 {    43 33 } g R^
   011 {    45  1 } RCL 1
   012 {       14 } yX
   013 {       20 } *
   014 {        1 } 1
   015 {       30 } -
   016 {    43 32 } g RTN

# End.
```
It looks like R0 contains 2, and R1 contains 1.  But what was in X and Y when you pressed SOLVE?

**Comment by jneb** [comment]
X and Y were both 0.5
But you seem to have found it already.
- Jurjen

Op di 16 jul 2024 04:48 schreef Bill Foote ***@***.***>:

**Comment by zathras** [comment]
> X and Y were both 0.5

Odd.  On the original simulator (version 2.1.14), when I  restore the state to what you give in the report, and I type "0.5 enter f solve 0", it finds the root.  No error.  So something's missing.  It would be nice if you could give instructions that work to reproduce the error, with the program you gave.  Can you do this?

**Comment by jneb** [comment]
For clarity I made a fresh debug copy (ignore the additional code).
Then I typed
1 STO 0 STO 1
.5 f SOLVE 0
and got error 8 (but it wasn't blinking, for some reason)
I checked if it matters if the delay is set to 50 ms instead of 0, but I
get the same result (but it takes forever, of course).
- Jurjen

Op di 16 jul 2024 16:03 schreef Bill Foote ***@***.***>:

**Comment by zathras** [comment]
> 1 STO 0 STO 1 .5 f SOLVE 0

In that case, you had 1 in the Y register, and 0.5 in the X register.  If you had wanted .5 in both X and Y, you would have had to have pressed enter before "f solve 0."

This is why we software developers ask for detailed, reproducible bug reports!

OK, doing that with v. 2.1.14, I get the blinking Error 8, as expected.

On the physical calculator, it gives "Error 0" (quite quickly, which is to be expected because the Y endpoint (1) generates a divide-by-zero.  Thanks.

**Comment by zathras** [comment]
Released in 2.1.15
